apps/src/lib/client/rpc.rs

```diff
@@ -1031,39 +1231,53 @@ pub async fn query_shielded_balance(
         (Some(token), false) => {
             // Compute the unique asset identifier from the token address
             let token = ctx.get(&token);
-            let asset_type = AssetType::new(
-                (token.clone(), epoch.0)
-                    .try_to_vec()
-                    .expect("token addresses should serialize")
-                    .as_ref(),
-            )
-            .unwrap();
-            let token_alias = lookup_alias(ctx, &token);
-            println!("Shielded Token {}:", token_alias);
             let mut found_any = false;
+            let token_alias = lookup_alias(ctx, &token);
+            println!(
+                "Shielded Token {}{}:",
+                token_alias,
+                args.sub_prefix
+                    .as_ref()
+                    .map(|k| format!("/{}", k))
+                    .unwrap_or_default()
+            );
             for fvk in viewing_keys {
-                // Query the multi-asset balance at the given spending key
-                let viewing_key = ExtendedFullViewingKey::from(fvk).fvk.vk;
-                let balance = if no_conversions {
-                    ctx.shielded
-                        .compute_shielded_balance(&viewing_key)
-                        .expect("context should contain viewing key")
-                } else {
-                    ctx.shielded
-                        .compute_exchanged_balance(
-                            client.clone(),
-                            &viewing_key,
-                            epoch,
-                        )
-                        .await
-                        .expect("context should contain viewing key")
-                };
-                if balance[&asset_type] != 0 {
-                    let asset_value =
-                        token::Amount::from(balance[&asset_type] as u64);
-                    println!("  {}, owned by {}", asset_value, fvk);
-                    found_any = true;
+                let mut balance = token::Amount::default();
+                for denom in MaspDenom::iter() {
```

Comment:
This logic seems to repeat what is at line 1049 of this file, maybe it could be factored out into a function? See also the comment at line 1070 of this file.